### PR TITLE
executor: rewrite "doRequestForDatums" to avoid close child during "Next()"

### PR DESCRIPTION
**Pull request by zz-jason**
to #4983

### CODE REVIEW

File: executor/distsql.go [hunk 1]
```diff
@@ -410,30 +410,33 @@ func handleIsExtra(col *expression.Column) bool {
...
 	// result returns one or more distsql.PartialResult and each PartialResult is returned by one region.
 	result        distsql.SelectResult
 	partialResult distsql.PartialResult
 	priority      int
-}
 
-// Schema implements the Executor Schema interface.
-func (e *TableReaderExecutor) Schema() *expression.Schema {
-	return e.schema
+	prepared   bool
```
Comment (coocood) [inline executor/distsql.go]: How about rename to `fetched`?

File: executor/distsql.go [hunk 2]
```diff
@@ -510,51 +535,52 @@ func (e *TableReaderExecutor) doRequestForHandles(goCtx goctx.Context, handles [
...
-	return e.schema
+	fetched    bool
+	request    *kv.Request
+	requestCtx goctx.Context
 }
 
 // Close implements the Executor Close interface.
 func (e *IndexReaderExecutor) Close() error {
+	if err := e.baseExecutor.Close(); err != nil {
+		return errors.Trace(err)
```
Comment (coocood) [inline executor/distsql.go]: Should continue close other resources.

**Comment by tiancaiamao** [comment]
Don't reuse the `ReaderExecutor` we can totally remove the 

```
doRequestForDatums(goCtx goctx.Context, datums [][]types.Datum) error
```

and simplify the logic a lot.

**Comment by zz-jason** [comment]
replaced by https://github.com/pingcap/tidb/pull/5031

**Pull request closed without merging**